### PR TITLE
Support ControlNet for Qwen-Image-Edit

### PR DESCRIPTION
**Pull request by dimitribarbot**
# What does this PR do?

Add ControlNet ([InstantX/Qwen-Image-ControlNet-Union](https://huggingface.co/InstantX/Qwen-Image-ControlNet-Union)) support for Qwen-Image-Edit. 

This pipeline enables two latent images to be used as inputs: one for Qwen-Image-Edit and another for Qwen-Image-ControlNet-Union. This provides greater control over the expected results.

## Inference

```python
import torch
from diffusers import QwenImageControlNetModel, QwenImageEditControlNetPipeline
from diffusers.utils import load_image

base_model = "Qwen/Qwen-Image-Edit"
controlnet_model = "InstantX/Qwen-Image-ControlNet-Union"

controlnet = QwenImageControlNetModel.from_pretrained(controlnet_model, torch_dtype=torch.bfloat16)

pipe = QwenImageEditControlNetPipeline.from_pretrained(
    base_model, controlnet=controlnet, torch_dtype=torch.bfloat16
).to("cuda")

image = load_image(
    "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/living_room.png"
).convert("RGB")
control_image = load_image(
    "https://huggingface.co/InstantX/Qwen-Image-ControlNet-Union/resolve/main/conds/depth.png"
)
prompt = (
    "Anime style of a swanky, minimalist living room with a huge floor-to-ceiling window letting in loads of natural light. "
    "A beige couch with white and beige cushions sits on a wooden floor, with a matching coffee table in front. "
    "The walls are a soft, warm beige, decorated with two framed botanical prints. A potted plant chills in the corner near the window. "
    "Sunlight pours through the leaves outside, casting cool shadows on the floor."
)
image = pipe(
    image=image,
    prompt=prompt,
    negative_prompt=" ",
    control_image=image,
    controlnet_conditioning_scale=1.5,
    width=control_image.size[0],
    height=control_image.size[1],
    num_inference_steps=30,
    true_cfg_scale=2.5,
).images[0]
image.save("qwenimage_edit_controlnet.png")
```

N.B.1. If this PR and image location are accepted, I will upload the `living_room.png` file to the `documentation-images` repository.
N.B.2. To achieve the desired result, set `controlnet_conditioning_scale` to a value greater than 1. A good starting point is 1.5.

### Depth

Input image:

living_room

Control image:

depth

```python
prompt = (
    "Anime style of a swanky, minimalist living room with a huge floor-to-ceiling window letting in loads of natural light. "
    "A beige couch with white and beige cushions sits on a wooden floor, with a matching coffee table in front. "
    "The walls are a soft, warm beige, decorated with two framed botanical prints. A potted plant chills in the corner near the window. "
    "Sunlight pours through the leaves outside, casting cool shadows on the floor."
)
```

Result:

living_room_edited

### Pose

Input image:

depth

Control image:

depth

```python
prompt = (
    "Make this man sit on a concrete ledge in front of a large circular window, with a cityscape reflected in the glass. "
    "The wall is cream-colored, and the sky is clear blue. His shadow is cast on the wall."
)
```

Result:

pose_with_controlnet

Whereas if we don't use controlnet:

pose_without_controlnet

N.B. All examples were created using the Nunchaku version of the transformer.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@yiyixuxu 
@asomoza

### CODE REVIEW

File: src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py [hunk 1]
```diff
@@ -0,0 +1,1105 @@
...
+        Enable sliced VAE decoding. When this option is enabled, the VAE will split the input tensor in slices to
+        compute decoding in several steps. This is useful to save some memory and allow larger batch sizes.
+        """
+        depr_message = f"Calling `enable_vae_slicing()` on a `{self.__class__.__name__}` is deprecated and this method will be removed in a future version. Please use `pipe.vae.enable_slicing()`."
+        deprecate(
+            "enable_vae_slicing",
+            "0.40.0",
+            depr_message,
+        )
+        self.vae.enable_slicing()
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py]:
```suggestion
```

File: src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py [hunk 2]
```diff
@@ -0,0 +1,1105 @@
...
+        Disable sliced VAE decoding. If `enable_vae_slicing` was previously enabled, this method will go back to
+        computing decoding in one step.
+        """
+        depr_message = f"Calling `disable_vae_slicing()` on a `{self.__class__.__name__}` is deprecated and this method will be removed in a future version. Please use `pipe.vae.disable_slicing()`."
+        deprecate(
+            "disable_vae_slicing",
+            "0.40.0",
+            depr_message,
+        )
+        self.vae.disable_slicing()
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py]:
```suggestion
```

File: src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py [hunk 3]
```diff
@@ -0,0 +1,1105 @@
...
+        compute decoding and encoding in several steps. This is useful for saving a large amount of memory and to allow
+        processing larger images.
+        """
+        depr_message = f"Calling `enable_vae_tiling()` on a `{self.__class__.__name__}` is deprecated and this method will be removed in a future version. Please use `pipe.vae.enable_tiling()`."
+        deprecate(
+            "enable_vae_tiling",
+            "0.40.0",
+            depr_message,
+        )
+        self.vae.enable_tiling()
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py]:
```suggestion
```

File: src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py [hunk 4]
```diff
@@ -0,0 +1,1105 @@
...
+        Disable tiled VAE decoding. If `enable_vae_tiling` was previously enabled, this method will go back to
+        computing decoding in one step.
+        """
+        depr_message = f"Calling `disable_vae_tiling()` on a `{self.__class__.__name__}` is deprecated and this method will be removed in a future version. Please use `pipe.vae.disable_tiling()`."
+        deprecate(
+            "disable_vae_tiling",
+            "0.40.0",
+            depr_message,
+        )
+        self.vae.disable_tiling()
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py]:
```suggestion
```

File: src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py [hunk 5]
```diff
@@ -0,0 +1,1105 @@
...
+        r"""
+        Enable sliced VAE decoding. When this option is enabled, the VAE will split the input tensor in slices to
+        compute decoding in several steps. This is useful to save some memory and allow larger batch sizes.
+        """
+        depr_message = f"Calling `enable_vae_slicing()` on a `{self.__class__.__name__}` is deprecated and this method will be removed in a future version. Please use `pipe.vae.enable_slicing()`."
+        deprecate(
+            "enable_vae_slicing",
+            "0.40.0",
+            depr_message,
+        )
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py]:
```suggestion
```

File: src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py [hunk 6]
```diff
@@ -0,0 +1,1105 @@
...
+        encoder_attention_mask = torch.stack(
+            [torch.cat([u, u.new_zeros(max_seq_len - u.size(0))]) for u in attn_mask_list]
+        )
+
+        prompt_embeds = prompt_embeds.to(dtype=dtype, device=device)
+
+        return prompt_embeds, encoder_attention_mask
+
+    def encode_prompt(
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py:300]: missing a `#Copied from` ?

File: src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py [hunk 7]
```diff
@@ -0,0 +1,1105 @@
...
+        return encoder_output.latent_dist.sample(generator)
+    elif hasattr(encoder_output, "latent_dist") and sample_mode == "argmax":
+        return encoder_output.latent_dist.mode()
+    elif hasattr(encoder_output, "latents"):
+        return encoder_output.latents
+    else:
+        raise AttributeError("Could not access latents of provided encoder_output")
+
+
+def calculate_dimensions(target_area, ratio):
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py:178]: Copied from?

File: src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py [hunk 8]
```diff
@@ -0,0 +1,1105 @@
...
+        width = 2 * (int(width) // (vae_scale_factor * 2))
+
+        latents = latents.view(batch_size, height // 2, width // 2, channels // 4, 2, 2)
+        latents = latents.permute(0, 3, 1, 4, 2, 5)
+
+        latents = latents.reshape(batch_size, channels // (2 * 2), 1, height, width)
+
+        return latents
+
+    def _encode_vae_image(self, image: torch.Tensor, generator: torch.Generator):
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit_controlnet.py:424]: copied from?

File: src/diffusers/models/transformers/transformer_qwenimage.py [hunk 9]
```diff
@@ -639,7 +639,9 @@ def forward(
             if controlnet_block_samples is not None:
                 interval_control = len(self.transformer_blocks) / len(controlnet_block_samples)
                 interval_control = int(np.ceil(interval_control))
-                hidden_states = hidden_states + controlnet_block_samples[index_block // interval_control]
+                sample = controlnet_block_samples[index_block // interval_control]
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_qwenimage.py]: can we adjust inputs in pipeline instead?

**Review by yiyixuxu — COMMENTED** [review]
thanks! i left some comments 
didn't know the instant x controlnet is compatible with qwen edit :)

**Comment by yiyixuxu** [comment]
hi @dimitribarbot 
about this, which feedback you are talking about? 
>However, this feedback also applies to the other Qwen-Image pipelines

if it is refering to the remove of `enable_vae_slicing` methods and such, that's because this is a new pipeline and we do not need to add a method that we already deprecated 

my review is not taking into consideration of this PR # https://github.com/huggingface/diffusers/pull/12322 at all, your PR will probably be merged first, and we can solve the conflicts from the other PR if there is any

File: src/diffusers/models/transformers/transformer_qwenimage.py (hunk 9 above)
Comment (dimitribarbot) [reply src/diffusers/models/transformers/transformer_qwenimage.py]: I added sample padding to the pipeline and rolled back that change. Let me know if that works for you.

**Comment by sayakpaul** [comment]
@dimitribarbot thanks for this PR. You can completely disregard https://github.com/huggingface/diffusers/pull/12322 for now. We're still brainstorming it (hence it's WIP).

**Comment by sayakpaul** [comment]
Thanks for your understanding and appreciate the efforts!
